### ISSUE TITLE
Identity Server 4 

### ISSUE THREAD

**Issue by keshavk** [dsuryd/dotNetify#101]
Can we use Identity Server 4 for Auth? If yes can I get example for this.

**Comment by dsuryd** [comment]
See issue #73.  The short answer is 'probably' because I don't use ID4, but a very similar library called [ASOS](https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/issues/350).  

Since this is in the periphery of this project's body of work, I won't write example.  But if you find the time to take it on, I encourage you to contribute back for the benefit of others with the same need.

**Comment by keshavk** [comment]
Thank you for reference, I have used ASOS and was trying to integrate ID4, And Yes, I will contribute back.

**Comment by hyounoo** [comment]
Can I get some help on integrating dotnetify with IDSV4 ?
Dotnetify fits perfect with my project except integrating with IDSV4.
Can I ask for an advice at least where to start from?
I've stuck at adding tokenValidationParameters shown in below example.
app.UseDotNetify(config =>
{
   var tokenValidationParameters = new TokenValidationParameters { /*...*/ }

   config.UseFilter<AuthorizeFilter>();
   config.UseJwtBearerAuthentication(tokenValidationParameters);
});

What should I add into TokenValidationParameters for IdentityServer4?

**Comment by keshavk** [comment]
Due to some other priority, I have not used IDS4. I have used ASOS. Following is the code which I have implemented.

config.UseJwtBearerAuthentication(new TokenValidationParameters
                {
                    IssuerSigningKey = new SymmetricSecurityKey(Encoding.ASCII.GetBytes(AuthServer.SecretKey)),
                    ValidateIssuerSigningKey = true,
                    ValidateAudience = false,
                    ValidateIssuer = false,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.FromSeconds(0)
                });
I know this will not going to help... but just for reference

**Comment by hyounoo** [comment]
I'v already examined your code inside

**Comment by hyounoo** [comment]
If there is no way to integrate id4 atm, Is it possible to use dotnetify viewmodel inside authorized api controller?

**Comment by dsuryd** [comment]
I do not know ID4 well enough to be able to instruct you, but since it's based on OAuth2 standard, I see no reason why integration wouldn't be possible.  I assume you're already familiar with ID4, hence your questions, so let me ask you this: how would you configure the ID4 to validate your JWT bearer tokens if you are just working on a regular ASP.NET Core Web API?

**Comment by hyounoo** [comment]
This is how I configure ASP.NET Core Web API
public void ConfigureServices(IServiceCollection services)
{
            services.AddAuthentication("Bearer")
                .AddIdentityServerAuthentication(options =>
                {
                    options.Authority = Configuration["IdentityServerAddress"];
                    options.RequireHttpsMetadata = false;

                    options.ApiName = "api1";
                });

            services.AddCors(options =>
            {
                // this defines a CORS policy called "default"
                options.AddPolicy("default", policy =>
                {
                    policy.WithOrigins("http://localhost:8080")
                        .AllowAnyHeader()
                        .AllowAnyMethod();
                });
            });
}

and on Configure
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
            ..........
            app.UseCors("default");
            app.UseAuthentication();
            ...........
}

and I can set [Authorize] attribute on api controllers

so basically there is no extra methods to validate JWT bearer tokens if you are using ID4.

I really hope to be able to use dotnetify with ID4 cause I really like how dotnetify simplifies integrating SignalR into real world examples.

I promise that I will contribute back with usage examples utilizing dotnetify with ID4 once I figure out how to combine them. Just need a little help to get me started.

**Comment by dsuryd** [comment]
IdentityServer4 will eventually create a `TokenValidationParameters` object to validate the token, so if you are going to send the token generated by their auth server from a dotNetify client, then it's a matter of figuring out what their default object is, which typically would specify the following information:
- the issuer issuing the token
- the issuer signing key to verify the token is indeed coming from the issuer
- audience of the token 
- whether to validate the issuer
- whether to validate the audience
- whether to validate the token lifetime

So looking at your configuration, I am going to guess that the default issuer is probably that `IdentityServerAddress` value from the configuration; I have no idea whether they sign it, and the audience is probably not validated.   

I would start by creating a new TokenValidationParameters object that does not validate anything, and run the token generated by your auth server into the validator method:

```
using System.IdentityModel.Tokens
...
var claimsPrincipal = new JwtSecurityTokenHandler().ValidateToken(testToken, testTokenValidationParameters, out SecurityToken validatedToken)
```

When things are working, then you can fill in the information in the parameters object and start enabling validation to make your app secure.

**Comment by hyounoo** [comment]
I have tried modifying AddAuthentcationServer method as below and had no luck.
your sample: services.AddAuthentication().AddOpenIdConnectServer(options =>{...}
to: services.AddAuthentication(options =>
            {
                options.DefaultScheme = "Cookies";
                options.DefaultChallengeScheme = "oidc";
            })
                   .AddCookie("Cookies")
                   .AddOpenIdConnect("oidc", options =>
                   {
                       options.SignInScheme = "Cookies";

                       options.Authority = "https://localhost:5000";
                       options.RequireHttpsMetadata = false;

                       options.TokenValidationParameters.NameClaimType = "name";

                       options.ClientId = "donetifychat"; // api client id
                       options.SaveTokens = true;
                   });

and then modified tokenValidationParameters as below.
var tokenValidationParameters = new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = false,
                    ValidateAudience = false,
                    ValidateIssuer = true,
                    ValidIssuer = "https://localhost:5000",
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.FromSeconds(0)
                };

I set ValidateIssuerSigningKey = false because my identityServer doesn't configure signing key.

when I signed in with identityserver with vue.js client and try to access securePageVM, I get below exception on return new JwtSecurityTokenHandler().ValidateToken(token, _tokenValidationParameters, out validatedToken);

IDX10501: Signature validation failed. Unable to match keys: 
kid: '[PII is hidden]', 
token: '[PII is hidden]'.

I'm am so sorry to keep bothering you but I'm in urgent situation to solve this project and willing to pay if you can spare some time to resolve this issue.

**Comment by dsuryd** [comment]
Your issue is with validating the token, so I think you need to get help from the IdentityServer4 forum.

**Comment by hyounoo** [comment]
Thanks for your input.
I've posted on stack overflow for this.
https://stackoverflow.com/questions/54500434/how-to-manually-authorize-identityserver4-implicit-granttype-with-dotnetify

I will comeback again hopefully:D

**Comment by kinetiq** [comment]
@hyounoo I'm trying to do something similar but with Auth0 instead of identity server. Any luck?

**Comment by dsuryd** [comment]
IdentityServer4 integration is now documented here: https://dotnetify.net/core/api/security
